daffodil-core/src/main/scala/org/apache/daffodil/core/dsom/AnnotatedSchemaComponent.scala

```diff
@@ -377,6 +378,118 @@ trait AnnotatedSchemaComponent
+   * and simple types. This is necessary because we have elements being referenced
+   * by multiple element ref, where the first element gets its propCache populated with
+   * the properties its used, and the second element doesn't because of sharing. With
+   * this val, we copy used properties from element refs to elements, group refs to groups,
+   * from elements to their simple types, if the simple type is annotated with the property,
+   * and from simple types to their base types
+   */
+  final lazy val propagateUsedProperties: Unit = {
+    this match {
+      case ref: GroupRef =>
+        val groupBeingReferenced = ref.groupDef
+        // if the groupBeingReferenced carries the property then add the property to its propCache
+        ref.propCache.foreach { kv =>
+          if (groupBeingReferenced.formatAnnotation.justThisOneProperties.contains(kv._1)) {
+            groupBeingReferenced.propCache.put(kv._1, kv._2)
+          }
+        }
+      case ref: AbstractElementRef =>
+        val elementBeingReferenced = ref.optReferredToComponent.get
+        ref.propCache.foreach(kv =>
+          // if the elementBeingReferenced or its simpleType carries the property
+          // then add the property to its propCache
+          if (
+            elementBeingReferenced.formatAnnotation.justThisOneProperties.contains(kv._1)
+            || elementBeingReferenced.optSimpleType.exists {
+              case std: SimpleTypeDefBase =>
+                std.formatAnnotation.justThisOneProperties.contains(kv._1)
+              case _ => false
+            }
+          ) {
+            // only add the used properties that the elementBeingReferenced or its simple type carries,
+            // not any local properties that aren't shared
+            elementBeingReferenced.propCache.put(kv._1, kv._2)
+            // propagate used properties from element to its simpletype and vice versa
+            elementBeingReferenced.propagateUsedProperties
+          }
+        )
+      case ele: ElementDeclMixin =>
+        val ost = ele.optSimpleType
+        ost.collect { case std: SimpleTypeDefBase =>
+          // if the type has used properties transfer it to the element that's using it
+          // we don't need the check for what it carries since all properties on the type are shared
+          // by the element using it
+          std.propCache.foreach(kv => ele.propCache.put(kv._1, kv._2))
+          // copy the used properties from the element to the type
+          ele.propCache.foreach(kv =>
+            if (std.formatAnnotation.justThisOneProperties.contains(kv._1)) {
+              // only add the used properties that the simple type carries, not any local element properties that aren't shared
+              std.propCache.put(kv._1, kv._2)
+            }
+          )
+          // propagate through the element's simple type base chain
+          std.propagateUsedProperties
```

Comment:
This propagation down the chain is the problematic part of this algorithm. One is potentially going to walk the length of this chain many more times than is necessary, and not by some small amount. 

Consider trying this:

Take the list of all annotated schema comopnents ASCs, and divide it up into lists of 

- all global group defs
- all global element defs

For all global group defs, process them by using Root.refMap to find all referencing global group refs and pull the propCache entries from there over onto the global group def. 

For all global element decls, process them by using Root.refMap to find all element refs and pull the propCache entries from there over onto the global element decl. 

For all global simple type defs, process them by using Root.refMap to find all _elements_ that reference them and pull the propCache entries from there onto the globalSimpleType def. 

For all _local_ simple type defs, process them by pulling the propCache entries from their enclosing element onto them. 

This leaves only global simple type defs with references to them from other simple type defs (which can be global or local). 

These need to be processed in a specific order to avoid redundant processing. Create a table of "done" global simple type defs, initially empty. 

Then start from all global simple type defs whose base is a primitive type (call these leaf simple type defs). 
Process each of these as follows.

Use Root.refMap to find the other simple type defs which reference this one. 

If they are found in the "done" table, pull their properties onto this simple type. 

If they are not found in the "done" table, recursively process them, (they will become "done") then pull their properties onto this simple type. 

When all the simple type defs which reference this one are "done" and their properties have been pulled onto this one, add this simple type def to the "done" table, and go on to the next in the list. 

That is a walk which visits all referencers first before any referenced, so will propagate the propCache entries visiting each simple type def exactly once.

Reply:
An additional thought: If you can walk in this "referencers before referenced" order by way of the Root.refMap, starting from the leaf simple types, the global group defs, and the global element defs (having complex type), then, for each ASC you can compute a separate value which is the set of all properties used. 

That is, you don't need to be doing put() calls on the propCache. You can read the propCache, but fill in a separate structure of just the properties that are used so as to isolate from those that are unused. This could be a separate structure which holds only properties in the referencer's propCache that are NOT found in the referenced object's propCache, so that this new structure is not redundant with the contents of the propCache. 

This would be cleaner and much more space efficient than reusing the propCache data structure in this way that reads it but also modifies it.

Reply:
> Root.refMap to find the other simple type defs which reference this one

Root.refMap doesn't track derived types to their base types, attempts to add it broke things and this comment in Nesting makes it clear it's not supposed to do that

```scala
   // The other kinds of references (to escape schemes, to named formats, or
   // from derived types to their base types) shouldn't be encountered here.
```